### ISSUE TITLE
DBus support

### ISSUE THREAD

**Issue by karim-ammouche** [copperspice/copperspice#226]
Hi,
There is a dbus directory in src/copperspice, but no CMakeLists.txt.
Is it an important effort to enable Dbus support?

**Comment by bgeller** [comment]
This library has not been fully migrated to CopperSpice however this is something which could be done. Some testing would be required.  Is this a library you need for a particular application?  Are you looking to communicate between two CopperSpice applications?

**Comment by karim-ammouche** [comment]
Yes, it is needed to replace a Qt application that communicates with
another console C++ application (not using Copperspice)
Do you think another module (e.g CsSvg) could be used as a starting point
for the porting? Does it require deep knowledge of Copperspice internals?

**Comment by bgeller** [comment]
Please correct me if I have misunderstood your questions.  It sounds like you are interested in working with us to get CsDBus working. This is fantastic. Yes, using another library like CsSvg will be a great start for creating the CsDBus cmake files.  

One of the next steps is to run PepperMill on this code which is not too involved and I would be happy to do this step for you. There may be some clean up in this area and it would not require too much CS internal knowledge. 

The next step is to adjust the Meta Object System in about ten classes. We can spend some time helping you fix the first one and then the others should be straight forward. 

We are committed to helping you through all of this. Please let us know how you would like to proceed. 

Barbara

**Comment by karim-ammouche** [comment]
Hi Barbara, I'm more familiar with C than C++ but if I can help that would be great! I have added CMakeLists.txt for dbus and run PepperMill. There are compilation errors because of missing header files (qmetatype.h, qobject_p.h). After removing the #includes, there are other errors:

src/dbus/qdbuscontext.cpp
include/QtCore/qglobal.h:1024:95: error: 'd_ptr' was not declared in this scope  
 1024 |     inline Class##Private* d_func() { return reinterpret_cast<Class##Private *>(qGetPtrHelper(d_ptr)); } \

src/dbus/qdbuscontext.cpp | 43 | 22 | error | 'class QObject' has no member named 'qt_metacast'

src/dbus/qdbusextratypes.h | 45 | 21 | error | 'QDBusObjectPath::QDBusObjectPath(const QString&)' cannot be overloaded with 'QDBusObjectPath::QDBusObjectPath(const QLatin1String&)'

src/dbus/qdbusextratypes.h | 51 | 14 | error | could not convert '*(const QDBusObjectPath*)this' from 'const QDBusObjectPath' to 'QString' {aka 'QString8'}

src/dbus/qdbusextratypes.h | 58 | 7 | error | type 'QString8' is not a direct base of 'QDBusObjectPath'

src/dbus/qdbusextratypes.h | 62 | 7 | error | type 'QString8' is not a direct base of 'QDBusObjectPath'

rc/dbus/qdbusextratypes.h | 65 | 8 | error | redefinition of 'QDBusObjectPath::QDBusObjectPath(const QString&)'

src/dbus/qdbusextratypes.h | 70 | 27 | error | cannot call member function 'CsString::CsBasicString<E, A>& CsString::CsBasicString<E, A>::operator=(const T&) [with T = QDBusObjectPath; E = CsString::utf8; A = std::allocator<unsigned char>]' without object

src/dbus/qdbusextratypes.h | 73 | 32 | error | cannot call member function 'QString8& QString8::operator=(const QString8&)' without object

src/dbus/qdbusextratypes.h | 95 | 21 | error | 'QDBusSignature::QDBusSignature(const QString&)' cannot be overloaded with 'QDBusSignature::QDBusSignature(const QLatin1String&)'

src/dbus/qdbusextratypes.h | 115 | 8 | error | redefinition of 'QDBusSignature::QDBusSignature(const QString&)'
...

The changes are visible here: https://github.com/karim-ammouche/copperspice

https://github.com/karim-ammouche/copperspice/commit/2845d4b22ec91e0040c8e8bf0cd7512869cb5d56
https://github.com/karim-ammouche/copperspice/commit/5ff3a487606fcac6e20ad4a2a36443f892da74ee
https://github.com/karim-ammouche/copperspice/commit/f1d1aee9dcb246384079801fbda3f7102f376219

**Comment by karim-ammouche** [comment]
String related errors were fixed in latest commit. There are other errors related to classes inheriting from QObjectPrivate. The class is not defined. Is there a method to port these classes to Copperspice?

**Comment by bgeller** [comment]
QObjectPrivate was indeed removed and all necessary data members were moved to QObject. Start by removing any includes or references to QObjectPrivate and then see what happens. We will help investigate each one as you get there.

We are rally excited about all of your progress.

**Comment by karim-ammouche** [comment]
> QObjeqMetaTypeIdctPrivate was indeed removed and all necessary data members were moved to QObject. Start by removing any includes or references to QObjectPrivate and then see what happens. We will help investigate each one as you get there.
> 
> We are rally excited about all of your progress.

Thanks for your support! Here are a few questions:
- Assuming a class inherits from QObjectPrivate, should it be changed to inherit from QObject instead? Or QObjectPrivate should be removed?
- In qdbuscontext.cpp, there is a call to obj->qt_metacast("QDBusContext"). qt_metacast is not defined. Can it be replaced by a cast to QDbusContext*? Or a more complex function needs to be written, like the one generated by the moc compiler?
- In qdbusmarshaller.cpp there is a call to arg.constData(), but constData() is not defined for QVariant. Does it mean the type T of arg must be known to call the corresponding arg<T>.getData() ?
- Replacing qRegisterMetaType<QDBusObjectPath>("QDBusObjectPath") by QVariant::typeToTypeId<QDBusObjectPath>() causes a static_assert failure: "Requested type name has not been registered" Does it mean CS_REGISTER_CLASS(QDBusObjectPath) must be added to csmeta.h ?

**Comment by bgeller** [comment]
These are all very good questions and a few will require a bit of research, however we can help you solve them. Taking a quick look, we have an answer for the first question. 

> Assuming a class inherits from QObjectPrivate, should it be changed to inherit from QObject instead? Or QObjectPrivate should be removed?

You need to remove the inheritance. Then add the following:

```
class QSomethingPrivate
{
  protected:
    QSomething *q_ptr;
       
  private:
    Q_DECLARE_PUBLIC(QSomething)
};          
```
 
 In all constructors for QSomething add the following line of code:
 
` d-Ptr->q_ptr = this;`

If you want to check out an example have a look at our code for **src/core/io/qsettings_p.h** and **src/core/io/qsettings.cpp**

If you need more info about this just let us know.

**Comment by bgeller** [comment]
(2) The code for `obj->qt_metacast("QDBusContext")` should change to a dynamtic_cast.  The next line of code will need to change as well since the "reinterpret_cast" is no longer necessary. Lines 43 - 45 can be replaced by the following: 

```
QDBusContext *q_ptr = dynamic_cast<QDBusContext>(obj);

if (q_ptr != nullptr) { 
 . . .
}
```

**Comment by karim-ammouche** [comment]
Hi Barbara, thanks for the information. Due to schedule constraint I couldn't work on the porting. In our use use case, using a dbus library and converting the types manually was a viable solution, because only simple types are exchanged (std::string, std::vector, etc.)

**Comment by bgeller** [comment]
Thank you for everything you have done so far. We will take what you have done and see how much more is required.  Maybe we still work together since you have a decent test case.
